### ISSUE TITLE
DBus Signals example

### ISSUE THREAD

**Issue by mharshe** [LEW21/pydbus#5]
I'm having a hard time wrapping my head around making a "signal receiver". All examples online use the old python-dbus library, and they seem to have the [bus.add_signal_receiver() method][sig]  . How does one go about implementing it in pydbus? 

(FWIW I'm trying to write the bluez test modules using pydbus, and the [bluez code][bluez] makes heavy use of the add_signal_receiver)

[sig]: http://dbus.freedesktop.org/doc/dbus-python/api/dbus.bus.BusConnection-class.html#add_signal_receiver
[bluez]: http://git.kernel.org/cgit/bluetooth/bluez.git/tree/test/test-discovery

**Comment by LEW21** [comment]
This is not currently supported in the pythonic API (until I find time to do it, which might be during the next 24 hours or next year - in the second case, patches welcome).

But you can do it quite easily with Gio's API:
```
# bus is a pydbus's SystemBus/SessionBus.
bus.con.signal_subscribe("sender_name", "iface_name", "signal_name", "object_path", None, 0, lambda *args: print(args))

from gi.repository import GObject
GObject.MainLoop().run()
```

Full documentation of signal_subscribe method is available at https://developer.gnome.org/gio/stable/GDBusConnection.html#g-dbus-connection-signal-subscribe

**Comment by LEW21** [comment]
API updated, as described in https://github.com/LEW21/pydbus#watch-for-new-systemd-jobs you can do:

```python
from pydbus import SystemBus
from gi.repository import GObject

bus = SystemBus()
systemd = bus.get(".systemd1")

systemd.JobNew.connect(print)
GObject.MainLoop().run()
```

Disconnecting is also supported:

```python
with systemd.JobNew.connect(print):
    GObject.MainLoop().run()
# Here the connection dies
````
or
```python
c = systemd.JobNew.connect(print):
GObject.MainLoop().run()
c.disconnect()
```

BTW I'm also thinking about adding a second way of connecting to signals, probably soon both will be supported:
```python
systemd = bus.get(".systemd1")

systemd.onJobNew = print
GObject.MainLoop().run()
systemd.onJobNew = None # disconnect
```

Then I'll release 0.3 to pypi.

**Comment by mharshe** [comment]
That looks great. I'll test this out soon to see how it works. I think the bluez stuff really needs more examples, and I think I'll be able to create some sample scripts using this. The official bluez sample scripts use the old python dbus a lot, so I have a reference to test this with multiple examples.

**Comment by polyphemus** [comment]
I was interested in using your library to listen signals send by dbus services, but I have a problem getting your examples to work.

When I try the first systemd example everything is fine until the `systemd.JobNew.connect(print)` line throws `AttributeError: 'CompositeObject' object has no attribute 'JobNew'`. It seems that the `systemd` object does not have the signal objects as members. If I check the ipython completion for members of the object I only the methods are listed.

The second example does not throw but will happily block on the run method and never print anything.

I'm running Arch Linux with dbus version 1.8.18 and systemd version 219.

**Comment by rrichardson** [comment]
I have the problem of missing JobNew as well.  It appears that only the "methods" property of the interface is being introspected and not the "signals" property.   

Also, I'm having a hard time loading the interface "org.freedesktop.DBus.Properties" which contains the PropertiesChanged signal, which is the one we need to watch to subscribe to changes to systemd daemons.

**Comment by rrichardson** [comment]
As per your suggestion, I went with : 
    
```python
    bus.con.signal_subscribe(None, "org.freedesktop.DBus.Properties", "PropertiesChanged", None, None, 0, lambda *args: print(args))
```

Which has worked swimmingly. It just takes 2 lines of code to set up a systemd daemon monitor (plus callback code) nice :)

**Comment by bewest** [comment]
The new signals interface is highly desirable, but only works when installing from source.  The demo is specific to systemd, which only relatively recently the new default (post sysV, upstart world).

**Comment by LEW21** [comment]
I've released 0.3 to pypi.
